Add the partial auth feature of Google account to the reconnection subpage

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Part of #1000

- Rename files in the *js/src/settings/reconnect-accounts* folder.
- Extract the `toScopeState` out as a util function and adjust some logic for checking the GLA required permission based on the current setup.
- Add the partial auth feature to the reconnection subpage

### Screenshots:

❗ Please note that these videos were recorded with local changes:
1. The same temporary hack as https://github.com/woocommerce/google-listings-and-ads/pull/1064#discussion_r736426466
2. Remove `hideIcon` props from all Google account card relevant components to show how it looks like after the #1065 is completed.

#### 🎥 . The original access check after reconnecting

💡  Even though accepts all permissions, asking user to disconnect is a priority action if user reconnects to an account that has no access to its previously connected GMC or Ads accounts.

https://user-images.githubusercontent.com/17420811/139220362-b018ddc3-3360-4f3e-b433-633a3db91566.mp4

#### 🎥 . Partial auth when granting without required permissions

https://user-images.githubusercontent.com/17420811/139225903-830687f9-8e5b-492a-b39b-8d44c7493a7d.mp4

#### 🎥 . If a user hasn't set up Google Ads account yet when reconnecting, then it could complete it without granting the ads permission
https://user-images.githubusercontent.com/17420811/139220392-a70276ff-8452-412c-b840-45a14b8fced7.mp4

### Detailed test instructions:

#### Test notes:

- 💡  The ads permission would be required if the Google Ads setup has been completed before entering reconnection process.
- 💡  Please ignore the problem of entering Google account choosing page first when going to Google OAuth page. It will be resolved once the dependency PR of #1063 in WCS is merged and deployed.

#### Pre-condition:

Use a Google account to complete the Setup MC flow. If the Setup MC flow has not been done, when you perform the following test step, the page will redirect you to the Setup MC page instead of reconnection page.

#### Test steps:

1. Disconnect Google account from the Connection Test page.
1. Go to a GLA page. It should redirect you to the reconnection page.
1. Connect with _another_ Google account with all/partial permissions, it should ask you to disconnect and reconnect due to the new connected account having no access to the previously linked Google Merchant Center account or Google Ads account. Click on the "Try another Google account" button to disconnect.
1. Connect with the original Google account without required permissions, it should re-ask authorization.
1. After re-authorizing the required permissions, it should redirect you to the dashboard page.

### Changelog entry

> Add - The partial auth feature of Google account to the reconnection subpage in the Settings page.

eason9487 committed:
Rename files in js/src/settings/reconnect-accounts

eason9487 committed:
Add index.js back to js/src/settings/reconnect-accounts/

eason9487 committed:
Extract toScopeState out as a util function

eason9487 committed:
Add reconnection mode supporting to GoogleAccountCard

eason9487 committed:
Add partial OAuth feature to the reconnection page

eason9487 committed:
Fix JS unit tests

eason9487 commented on js/src/utils/toScopeState.js:
```diff
@@ -0,0 +1,47 @@
+		scopes.includes( SCOPE.CONTENT ) &&
+		scopes.includes( SCOPE.SITE_VERIFICATION_VERIFY_ONLY );
+
+	state.glaRequired = adsSetupComplete
+		? state.gmcRequired && state.adsRequired
+		: state.gmcRequired;
```
📝  I extracted `toScopeState` from *js/src/hooks/useGoogleAccount.js* and changed a logic. The primary part is that replaced the `allRequired` with `glaRequired`. More details can be found [here](https://github.com/woocommerce/google-listings-and-ads/commit/2838b686779ceeee7a556310ab628191b8d9c5d4#diff-0c53f82583bf8825ea0072572f6678bb21b3b7d5bdcc2f93af1246f66781a4d9).

ecgan replied:
I was having a bit of difficulty trying to understand why we need `state.glaRequired` (and yes I read the JSDoc comment about it: `Whether has the required scopes of GLA plugin based on the current GLA setup`).

In the end I finally got to understand it from the PR description and video: "If a user hasn't set up Google Ads account yet when reconnecting, then it could complete it without granting the ads permission". 😄 

❔ Perhaps we can elaborate more on the meaning of `"current GLA setup"` in the JSDoc.

eason9487 replied:
Thanks for pointing out the unclear parts. 👍  Added description by https://github.com/woocommerce/google-listings-and-ads/pull/1067/commits/abaa7a567d51cccaf93b925f89e2ed857c9c1804.

ecgan commented on js/src/components/google-account-card/google-account-card.js:
```diff
@@ -16,15 +17,17 @@ export default function GoogleAccountCard( { disabled = false } ) {
 		return <ConnectedGoogleAccountCard googleAccount={ google } />;
 	}
 
+	const nextPageName = glaData.mcSetupComplete ? 'reconnect' : 'setup-mc';
```
❔ `nextPageName` depends only on `glaData.mcSetupComplete`. Instead of defining it here in `GoogleAccountCard` and then pass it down to `AuthorizeGoogleAccountCard` (via `nextPageName` props) and then `useGoogleAuthorization`, we could probably just define it inside `useGoogleAuthorization`? Also, `nextPageName` props is not being used elsewhere, so we don't really need that prop? 🤔

eason9487 replied:
> we could probably just define it inside `useGoogleAuthorization`?

The `nextPageName` will be another value that doesn't depend on `glaData.mcSetupComplete` while checking the permission on the Google Ads setup. So we have to keep it as a parameter of `useGoogleAuthorization`.

> Also, `nextPageName` props is not being used elsewhere, so we don't really need that prop?

You're right. 👍  I moved it into the `<AuthorizeGoogleAccountCard>` component as an internal variable by https://github.com/woocommerce/google-listings-and-ads/pull/1067/commits/3805e720dfd310f2dd6c9e8fc8f2d6cf28e61b24. Also, I renamed it to `pageName` due to I will use it for the tracking event properties as well. I think `pageName` naming should be less confusing.

eason9487 replied:
> I renamed it to `pageName` due to I will use it for the tracking event properties as well

Please refer to this commit https://github.com/woocommerce/google-listings-and-ads/pull/1069/commits/c4761c141286e19c937f41e89036871021ad30d6 for more details.

ecgan commented on js/src/utils/toScopeState.js:
```diff
@@ -0,0 +1,47 @@
+ */
+export default function toScopeState(
+	scopes = [],
+	adsSetupComplete = glaData.adsSetupComplete
```
❔ We don't really need this parameter? 🤔

eason9487 replied:
Yes, this parameter is not needed for implementation.

This parameter is there for the convenience of not having to import `glaData` from constants.js and pass it from the consumer side during implementation. In addition, for the sake of keeping the util function as pure as possible, and not having to specifically mock `glaData` when writing tests later.

ecgan replied:
> not having to specifically mock `glaData` when writing tests later.

From the way I see it, tests should reflect requirements and how the function is used in the application code.

In our application code, we use `toScopeState (arg1)`, we don't have any usage of `toScopeState (arg1, arg2)`. 

So in the test code, we should write tests with `toScopeState (arg1)` usage. Yes, we would have to mock `glaData`, and I'd say that is expected, because it reflects how it is being used in application code. 

If we were to write tests with `toScopeState (arg1, arg2)` usage, someone (my future self? 🙈  ) doing a "find all references" on `toScopeState` will wonder: 

- "Why do we have tests for `toScopeState (arg1, arg2)` that is not being used in application?"
- "Why do we not have tests for `toScopeState (arg1)` that is being used in application?"

Is my concern above valid? 🤔

eason9487 replied:
This pattern comes from my past experience, but it's only a team/company level convention. I'd like to adjust it if it brings ambiguous meaning. Will open a PR for it later and along with adding its test cases.

eason9487 replied:
PR opened #1071

ecgan reviewed (approved):
Tested, working as expected (I added a pre-condition section about completing Setup MC flow in the test instruction in the PR description). Left three minor comments below. Approving this as there is nothing major blocking us. 👍

eason9487 committed:
Add description for "current GLA setup" in the toScopeState function.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1067#discussion_r738432485

eason9487 committed:
Move nextPageName as an internal variable instead of a component prop.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1067#discussion_r738411808
